### PR TITLE
Add document question answering examples

### PR DESCRIPTION
**Pull request by sudeepag**
This PR adds two examples for document question answering using Langchain.

- The first example, `retrieval_qa_sotu_example`, loads in the State of the Union speech and answers 5 questions from it. It uses retrieval QA, which works by chunking the document, getting the embeddings for each chunk, and performing a similarity search to find the chunks most similar to the prompt.

[Screencast from 08-29-2023 01:02:24 AM.webm](https://github.com/mlc-ai/mlc-llm/assets/11940172/f6d6b07d-575b-4fc0-82d5-de28a7c55e8b)

- The second example, `retrieval_qa_mlc_docs_example`, uses the [MLC documentation](https://mlc.ai/mlc-llm/docs/) to answer some questions about the MLC project and functionalities. Similar to the above demo, we load in the `rst` files into a vector store, and perform an embedding search to find similar chunks of the documentation. In this demo, the question answering is done interactively, so the user can ask any question they would like.

[Screencast from 08-29-2023 01:06:37 AM.webm](https://github.com/mlc-ai/mlc-llm/assets/11940172/a8225a01-c88e-4492-9a33-3c36720a1d7e)

### CODE REVIEW

File: mlc_llm/core.py [hunk 1]
```diff
@@ -359,16 +359,16 @@ def mod_transform_before_build(
...
             max_seq_len = args.max_seq_len
-        elif "max_sequence_length" in config:
-            max_seq_len = config["max_sequence_length"]
+        elif config.max_sequence_length:
+            max_seq_len = config.max_sequence_length
 
         if max_seq_len:
             mod = fuse_split_rotary_embedding(
-                mod, config["num_attention_heads"], config["hidden_size"], max_seq_len
+                mod, config.num_attention_heads, config.hidden_size, max_seq_len
```
Comment (sudeepag) [inline mlc_llm/core.py:371]: @MasterJH5574 @nverke Could you please confirm that this fixes the bug outlined in https://github.com/mlc-ai/mlc-llm/pull/806#issuecomment-1696766685?

File: examples/rest/python/sample_langchain.py [hunk 2]
```diff
@@ -2,9 +2,19 @@
...
 from langchain.chains.question_answering import load_qa_chain
 from langchain.llms import OpenAI
+from langchain.text_splitter import CharacterTextSplitter
+from langchain.chains import RetrievalQA
+from langchain.vectorstores import Chroma
+
+# Note that Langchain support for embedding documents using MLC is currently blocked on
+# https://github.com/langchain-ai/langchain/pull/7815
+# We have subclassed `OpenAIEmbeddings` in the meantime to get around this dependency.
+from mlc_chat.embeddings.openai import MLCEmbeddings
```
Comment (sudeepag) [inline examples/rest/python/sample_langchain.py:15]: @tqchen Subclassed `OpenAIEmbeddings` with a new class `MLCEmbeddings`, lmk if this is what you had in mind.

File: mlc_llm/core.py (hunk 1 above)
Comment (nverke) [reply mlc_llm/core.py:371]: I can test this out tomorrow! Must have accidentally tested the wrong branch 🤦‍♂️

Comment (sudeepag) [reply mlc_llm/core.py:371]: Thank you!

Comment (masahi) [reply mlc_llm/core.py:371]: @sudeepag Sorry we keep making bugs with this issue. Can you make a separate PR for this change to merge this quickly?

Comment (MasterJH5574) [reply mlc_llm/core.py:371]: I think #826 fixes the issue. Could you try drop all changes here and rebase on to the latest main?

**Review by tqchen — APPROVED** [review]

File: mlc_llm/core.py [hunk 3]
```diff
@@ -359,16 +359,16 @@ def mod_transform_before_build(
...
     )  # pylint: disable=not-callable
 
-    if "num_attention_heads" in config and "hidden_size" in config:
+    if config.num_attention_heads and config.hidden_size:
         max_seq_len = None
         if args.max_seq_len > 0:
             max_seq_len = args.max_seq_len
-        elif "max_sequence_length" in config:
-            max_seq_len = config["max_sequence_length"]
+        elif config.max_sequence_length:
```
Comment (MasterJH5574) [inline mlc_llm/core.py:366]: > I think #826 fixes the issue. Could you try drop all changes here and rebase on to the latest main?

It turns out that #826 is not a complete fix as found by @CharlieFRuan. This line need to updated with `hasattr` as well.

Comment (CharlieFRuan) [reply mlc_llm/core.py:366]: Sorry I did not see this PR and made a separate one: https://github.com/mlc-ai/mlc-llm/pull/829.

I'll close that one as @sudeepag's should fix the issue.

File: mlc_llm/core.py (hunk 1 above)
Comment (nverke) [reply mlc_llm/core.py:371]: Created https://github.com/mlc-ai/mlc-llm/pull/831 to add a test for this and resolve the error that is handled here.

Comment (nverke) [reply mlc_llm/core.py:371]: The hasattr is not enough because the objects are no longer dicts and therefore cannot be subscripted.

**Pull request merged**